### PR TITLE
Upload reports to S3 + Gzip

### PR DESCRIPTION
**Pull request by oilnam**
Since the size of a report can approach 1MB and we want to store up to 10 reports/hour, I thought it would be good to gzip them. Sadly this only shrink them at ~50% of their original size, but still better than nothing.

### CODE REVIEW

File: src/main/scala/com/gu/contentapi/services/S3.scala [hunk 1]
```diff
@@ -1,9 +1,64 @@
...
+    new InstanceProfileCredentialsProvider()
+  )
+
+  val reportsBucketName = "apple-news-analytics/reports"
 
 }
+
+object Compress {
+
+  def gzip(content: String): ByteArrayOutputStream = {
```
Comment (cb372) [inline src/main/scala/com/gu/contentapi/services/S3.scala]: I would return an `Array[Byte]` from this method. Returning streams from functions is best avoided if possible, as the caller is not sure what state the stream is in. Should they close it when they're done with it? Is it already closed? etc.

File: src/main/scala/com/gu/contentapi/services/S3.scala [hunk 2]
```diff
@@ -1,9 +1,64 @@
...
+import com.amazonaws.services.s3.model.{ ObjectMetadata, PutObjectResult }
+import com.typesafe.scalalogging.LazyLogging
+import org.joda.time.DateTime
+
+import scala.util.{ Try, Success, Failure }
+
+trait S3 extends LazyLogging {
+  import S3._
+
+  val s3: AmazonS3Client = new AmazonS3Client(credentialsProvider).withRegion(Regions.EU_WEST_1)
```
Comment (cb372) [inline src/main/scala/com/gu/contentapi/services/S3.scala]: I would move this into the `S3` companion object, so that it becomes a singleton. If multiple classes extend their trait, each one will get its own instance of `AmazonS3Client`, which is probably not what you want.

To be honest, I would get rid of the whole trait and just have `object S3`.

File: src/main/scala/com/gu/contentapi/services/S3.scala [hunk 3]
```diff
@@ -1,9 +1,64 @@
...
+
+  def uploadToS3(content: String): Try[PutObjectResult] = {
+    val key = s"${DateTime.now}.gz"
```
Comment (cb372) [inline src/main/scala/com/gu/contentapi/services/S3.scala]: Do you know the exact time that the report was created? If so, it might be better to use that for the filename.

File: src/main/scala/com/gu/contentapi/services/S3.scala [hunk 4]
```diff
@@ -2,30 +2,37 @@ package com.gu.contentapi.services
...
-
 import scala.util.{ Try, Success, Failure }
 
-trait S3 extends LazyLogging {
-  import S3._
+object S3 extends LazyLogging {
 
-  val s3: AmazonS3Client = new AmazonS3Client(credentialsProvider).withRegion(Regions.EU_WEST_1)
+  val credentialsProvider = new AWSCredentialsProviderChain(
+    new EnvironmentVariableCredentialsProvider(),
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/contentapi/services/S3.scala]: This isn't needed.

File: src/main/scala/com/gu/contentapi/services/S3.scala [hunk 5]
```diff
@@ -2,30 +2,37 @@ package com.gu.contentapi.services
...
+    new SystemPropertiesCredentialsProvider(),
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/contentapi/services/S3.scala]: Or this.

Comment (LATaylor-guardian) [inline src/main/scala/com/gu/contentapi/services/S3.scala]: Or this.

File: src/main/scala/com/gu/contentapi/services/S3.scala [hunk 6]
```diff
@@ -2,30 +2,37 @@ package com.gu.contentapi.services
...
+    new ProfileCredentialsProvider("capi"),
+    new ProfileCredentialsProvider(),
+    new InstanceProfileCredentialsProvider()
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/contentapi/services/S3.scala]: Or this.

File: src/main/scala/com/gu/contentapi/services/S3.scala [hunk 7]
```diff
@@ -2,30 +2,37 @@ package com.gu.contentapi.services
...
 
-trait S3 extends LazyLogging {
-  import S3._
+object S3 extends LazyLogging {
 
-  val s3: AmazonS3Client = new AmazonS3Client(credentialsProvider).withRegion(Regions.EU_WEST_1)
+  val credentialsProvider = new AWSCredentialsProviderChain(
+    new EnvironmentVariableCredentialsProvider(),
+    new SystemPropertiesCredentialsProvider(),
+    new ProfileCredentialsProvider("capi"),
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/contentapi/services/S3.scala]: This is only needed locally - when running the Lambda in AWS it will not use it.

File: src/main/scala/com/gu/contentapi/services/S3.scala [hunk 8]
```diff
@@ -2,30 +2,37 @@ package com.gu.contentapi.services
...
+    new ProfileCredentialsProvider(),
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/contentapi/services/S3.scala]: This isn't needed.

File: src/main/scala/com/gu/contentapi/services/S3.scala [hunk 9]
```diff
@@ -2,30 +2,37 @@ package com.gu.contentapi.services
...
+    new InstanceProfileCredentialsProvider()
+  )
+
+  val reportsBucketName = "apple-news-analytics/reports"
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/contentapi/services/S3.scala]: This is a constant, so `reportsBucketName` can be `ReportsBucketName `

File: src/main/scala/com/gu/contentapi/services/S3.scala [hunk 10]
```diff
@@ -2,30 +2,37 @@ package com.gu.contentapi.services
...
+
+  val client: AmazonS3Client = new AmazonS3Client(credentialsProvider).withRegion(Regions.EU_WEST_1)
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/contentapi/services/S3.scala]: Region should be configurable via configuration.

File: src/main/scala/com/gu/contentapi/Lambda.scala [hunk 11]
```diff
@@ -17,15 +21,25 @@ class Lambda extends LazyLogging {
...
     }
 
   }
 
-  // TODO: do something useful with this report
-  private def downloadReport(url: String): Unit = {
+  private def downloadReport(successfulReport: SuccessfulReport): Unit = {
     logger.info("Downloading report...")
-    println(Http(url).asString.body)
+    val report = Http(successfulReport.link).asString.body
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/contentapi/Lambda.scala]: Do you need to check the response code of the request here? What if it's not 200, will you end up uploading something that isn't the report, but rather garbage?

File: src/main/scala/com/gu/contentapi/services/S3.scala [hunk 12]
```diff
@@ -1,9 +1,54 @@
...
+object S3 extends LazyLogging {
 
-  lazy val s3: AmazonS3Client = new AmazonS3Client()
+  val credentialsProvider = new AWSCredentialsProviderChain(
+    new ProfileCredentialsProvider("capi")
+  )
+
+  private val ReportsBucketName = "apple-news-analytics/reports"
+
+  val client: AmazonS3Client = new AmazonS3Client(credentialsProvider).withRegion(Regions.EU_WEST_1)
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/contentapi/services/S3.scala]: Region should be configurable like all of our applications.

File: src/main/scala/com/gu/contentapi/client/Reports.scala [hunk 13]
```diff
@@ -3,14 +3,22 @@ package com.gu.contentapi.client
...
 import com.typesafe.scalalogging.LazyLogging
+import io.circe.{ DecodingFailure, HCursor, Decoder, Json }
+import io.circe.jawn._
+import org.joda.time.DateTime
 import scalaj.http.Http
 
 class Reports extends LazyLogging {
   import Reports._
 
+  implicit val dateDecoder = new Decoder[DateTime] {
```
Comment (LATaylor-guardian) [inline src/main/scala/com/gu/contentapi/client/Reports.scala]: I think this should live inside of the companion object? @cb372

Comment (cb372) [inline src/main/scala/com/gu/contentapi/client/Reports.scala]: Why is Reports a class anyway? It doesn't take any constructor params, so let's make it an object.

Comment (LATaylor-guardian) [inline src/main/scala/com/gu/contentapi/client/Reports.scala]: Touche.

File: src/main/scala/com/gu/contentapi/Lambda.scala [hunk 14]
```diff
@@ -2,30 +2,64 @@ package com.gu.contentapi
...
+class Lambda extends LazyLogging with Retry {
 
   def generateReport(): Unit = {
 
-    val appleClient = new Reports
-
     val reportLink = for {
-      reportId <- appleClient.createArticleReport(Config.channelIdUk)
-      reportLink <- appleClient.readReportStatus(reportId)
+      reportId <- Reports.createArticleReport(Config.channelIdUk)
```
Comment (cb372) [inline src/main/scala/com/gu/contentapi/Lambda.scala]: Does this mean that the report only contains pageview data for the UK channel? Do we need to get reports for all 3 channels and sum the pageviews?

File: src/main/scala/com/gu/contentapi/Lambda.scala [hunk 15]
```diff
@@ -2,30 +2,64 @@ package com.gu.contentapi
...
+        None
+      }
+    }
   }
+
+}
+
+trait Retry {
+
+  def retry[R](retries: Int)(fn: => Option[R]): Option[R] = {
```
Comment (cb372) [inline src/main/scala/com/gu/contentapi/Lambda.scala]: Beautiful ❤️

Comment (LATaylor-guardian) [inline src/main/scala/com/gu/contentapi/Lambda.scala]: What a function signature!!

File: src/main/scala/com/gu/contentapi/Lambda.scala (hunk 14 above)
Comment (oilnam) [inline src/main/scala/com/gu/contentapi/Lambda.scala]: Yes, but I'd like to do that in a separate PR :)

**Comment by cb372** [comment]
👍
